src/garage/_dtypes.py

```diff
@@ -167,3 +167,199 @@ def __new__(cls, env_spec, observations, actions, rewards, terminals,
+            'env_info',
+            'agent_info',
+        ])):
+    # pylint: disable=missing-return-doc, missing-return-type-doc, missing-param-doc, missing-type-doc  # noqa: E501
+    r"""A tuple representing a batch of whole trajectories.
+
+    A :class:`Sample` represents single sample sampled from the environment.
+
+    Attributes:
+        env_spec (garage.envs.EnvSpec): Specification for the environment from
+            which this data was sampled.
+        observation (float): Observation at the current time step in the
+            environment.
+        action (float): Action sampled from the policy given the observation
+            at the current time step.
+        reward (float): The reward for taking the action at the current state.
+        terminals (boolean): A boolean representing the termination signal for
+            the current time step.
+        env_info (dict): A dict containing the env_info returned from the
+            environment. For example, this may contain whether the termination
+            of the environment is due to the time step limit expiration or a
+            terminal state.
+        agent_info (dict): A dict of containing the agent_info returned by the
+            agent/policy. For example, this may contain the hidden states from 
+            an RNN policy.
+
+
+    Raises:
+        ValueError: If any of the above attributes do not conform to their
+            prescribed types and shapes.
+
+    """
+    __slots__ = ()
+
+    def __new__(cls, env_spec, observation, action, reward, next_observation,
+                terminal, env_info, agent_info):  # noqa: D102
+        # pylint: disable=too-many-branches
+
+        if type(observation) != float:
+            raise ValueError(
+                'observation must be type {}, but got type {} '
+                'instead.'.format(float, type(observation)))
+        # observation
+        if not env_spec.observation_space.contains(observation):
+            raise ValueError(
+                'observation must conform to observation_space {}, but got '
+                'data with shape {} instead.'.format(
+                    env_spec.observation_space, observation))
+
+        if type(next_observation) != float:
+            raise ValueError(
+                'next_observation must be type {}, but got type {} '
+                'instead.'.format(float, type(next_observation)))
+        # next_observation
+        if not env_spec.observation_space.contains(next_observation):
+            raise ValueError(
+                'next_observation must conform to observation_space {}, but got '
+                'data with shape {} instead.'.format(
+                    env_spec.observation_space, next_observation))
+
+        # action
+        # TODO: Shape check can be removed for gym>=0.15.4 # pylint: disable=fixme # noqa: E501
+        if not (env_spec.action_space.contains(action)
+                and action.shape == env_spec.action_space.shape):
```

Comment:
I see this is copied from the old code, but the shape test is actually redundant.